### PR TITLE
Use C++11 chrono instead prehistoric homegrown stuff.

### PR DESCRIPTION
**Pull request by baldersheim**
@havardpe PR

### CODE REVIEW

File: searchcore/src/vespa/searchcore/proton/server/transactionlogmanagerbase.h [hunk 1]
```diff
@@ -5,25 +5,27 @@
...
+    TransLogClient::Session::UP     _tlcSession;
+    vespalib::string                _domainName;
+    mutable std::mutex              _replayLock;
     mutable std::condition_variable _replayCond;
-    volatile bool               _replayDone;
-    bool                        _replayStarted;
-    double                      _replayStartTime;
+    volatile bool                   _replayDone;
+    bool                            _replayStarted;
+    mutable fastos::StopWatch       _replayStopWatch;
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/server/transactionlogmanagerbase.h]: prefer not using mutable when not needed

File: searchcore/src/vespa/searchcore/proton/server/transactionlogmanagerbase.h [hunk 2]
```diff
@@ -36,8 +38,7 @@ class TransactionLogManagerBase {
     StatusResult init();
 
     void internalStartReplay();
-    virtual void doLogReplayComplete(const vespalib::string &domainName,
-                                     int64_t elapsedTime) const = 0;
+    virtual void doLogReplayComplete(const vespalib::string &domainName, int64_t elapsedTime) const = 0;
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/server/transactionlogmanagerbase.h]: consider using duration for elapsed time

**Review by havardpe — COMMENTED** [review]
I suggest changing the StopWatch to sampling the time in the constructor and each time elapsed is called, make elapsed const and remove stop.

Timed wait loops can be further simplified when using std::condition_variable since we can then do wait_until.

chrono time_point/duration should extend their reach to replace generic double/int64_t time variables.

**Comment by havardpe** [comment]
please make a new pull request for additional changes

File: searchlib/src/vespa/searchlib/features/random_normal_feature.cpp [hunk 3]
```diff
@@ -64,14 +62,14 @@ RandomNormalBlueprint::setup(const search::fef::IIndexEnvironment & env,
...
+
+fef::FeatureExecutor &
+RandomNormalBlueprint::createExecutor(const fef::IQueryEnvironment &, vespalib::Stash &stash) const
 {
     uint64_t seed = _seed;
     if (seed == 0) {
-        FastOS_Time time;
-        time.SetNow();
-        seed = static_cast<uint64_t>(time.MicroSecs()) ^
+        seed = static_cast<uint64_t>(duration_cast<microseconds>(steady_clock::now().time_since_epoch()).count()) ^
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/features/random_normal_feature.cpp]: should use actual time for random seed

File: searchlib/src/vespa/searchlib/features/randomfeature.cpp [hunk 4]
```diff
@@ -63,14 +59,14 @@ RandomBlueprint::setup(const search::fef::IIndexEnvironment & env,
...
+
+fef::FeatureExecutor &
+RandomBlueprint::createExecutor(const fef::IQueryEnvironment &env, vespalib::Stash &stash) const
 {
     uint64_t seed = _seed;
     if (seed == 0) {
-        FastOS_Time time;
-        time.SetNow();
-        seed = static_cast<uint64_t>(time.MicroSecs()) ^
+        seed = static_cast<uint64_t>(duration_cast<microseconds>(steady_clock::now().time_since_epoch()).count()) ^
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/features/randomfeature.cpp]: should use actual time for seed

File: slobrok/src/vespa/slobrok/server/reserved_name.cpp [hunk 5]
```diff
@@ -1,2 +1,23 @@
...
+      _reservedTime(steady_clock::now()),
+      isLocal(local)
+{ }
+
+bool
+ReservedName::stillReserved() const {
+    return (milliseconds() < 15000);
+}
+
+int64_t ReservedName::milliseconds() const {
```
Comment (havardpe) [inline slobrok/src/vespa/slobrok/server/reserved_name.cpp]: seconds is not milliseconds...

File: searchlib/src/vespa/searchlib/features/debug_wait.cpp [hunk 6]
```diff
@@ -22,29 +19,19 @@ class DebugWaitExecutor : public FeatureExecutor
...
-        if (_params.busyWait) {
-            for (int i = 0; i < 1000; i++)
-                ;
-        } else {
-            int rem = (int)(millis - time.MilliSecsToNow());
-            std::this_thread::sleep_for(std::chrono::milliseconds(rem));
-        }
-    }
-    outputs().set_number(0, 1.0e-6 * time.MicroSecsToNow());
+    steady_clock::time_point start = steady_clock::now();
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/features/debug_wait.cpp]: this time interval measurement could use StopWatch

File: searchlib/src/vespa/searchlib/features/debug_wait.cpp [hunk 7]
```diff
@@ -22,29 +19,19 @@ class DebugWaitExecutor : public FeatureExecutor
...
+    fastos::StopWatch::waitAtLeast(microseconds(static_cast<long>(_params.waitTime * 1000000)), _params.busyWait);
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/features/debug_wait.cpp]: this waiting has nothing to do with a StopWatch

File: messagebus/src/vespa/messagebus/message.cpp [hunk 8]
```diff
@@ -55,23 +56,21 @@ Message::swapState(Routable &rhs)
...
+    _timeReceived = steady_clock::now();
     return *this;
 }
 
-Message &
-Message::setTimeReceivedNow()
-{
-    _timeReceived.SetNow();
-    return *this;
+uint64_t
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/message.cpp]: Done

File: messagebus/src/vespa/messagebus/message.h [hunk 9]
```diff
@@ -13,9 +13,10 @@ namespace mbus {
  */
 class Message : public Routable {
 private:
+    using time_point = std::chrono::steady_clock::time_point;
     Route       _route;
-    FastOS_Time _timeReceived;
-    uint64_t    _timeRemaining;
+    time_point  _timeReceived;
+    int64_t     _timeRemaining;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/message.h]: time_point and milliseconds it is.

File: messagebus/src/vespa/messagebus/routablequeue.cpp [hunk 10]
```diff
@@ -39,15 +41,14 @@ RoutableQueue::enqueue(Routable::UP r)
 Routable::UP
 RoutableQueue::dequeue(uint32_t msTimeout)
 {
-    FastOS_Time t;
-    t.SetNow();
-    uint32_t msLeft = msTimeout;
+    steady_clock::time_point startTime = steady_clock::now();
+    uint64_t msLeft = msTimeout;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/routablequeue.cpp]: Simplified

File: messagebus/src/vespa/messagebus/routing/resender.cpp [hunk 11]
```diff
@@ -87,7 +83,7 @@ Resender::scheduleRetry(RoutingNode &node)
         TraceLevel::COMPONENT,
         vespalib::make_string("Message scheduled for retry %u in %.3f seconds.", retry, delay));
     msg.setRetry(retry);
-    _queue.push(Entry((uint64_t)(_time.MilliSecsToNow() + delay * 1000), &node));
+    _queue.push(Entry(steady_clock::now() + milliseconds(static_cast<long>(delay * 1000)), &node));
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/routing/resender.cpp]: Simplified.

File: slobrok/src/vespa/slobrok/server/reserved_name.cpp (hunk 5 above)
Comment (baldersheim) [reply slobrok/src/vespa/slobrok/server/reserved_name.cpp]: correct....ed

File: messagebus/src/vespa/messagebus/message.cpp [hunk 12]
```diff
@@ -55,23 +56,21 @@ Message::swapState(Routable &rhs)
...
+Message::getTimeReceived() const {
+    return duration_cast<milliseconds>(_timeReceived.time_since_epoch()).count();
 }
 
 uint64_t
 Message::getTimeRemainingNow() const
 {
-    return (uint64_t)std::max(0.0, _timeRemaining - _timeReceived.MilliSecsToNow());
+    return std::max(0L, _timeRemaining - duration_cast<milliseconds>(steady_clock::now() - _timeReceived).count());
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/message.cpp]: Simplified

File: messagebus/src/vespa/messagebus/systemtimer.cpp [hunk 13]
```diff
@@ -1,15 +1,14 @@
...
+using namespace std::chrono;
 namespace mbus {
 
 uint64_t
 SystemTimer::getMilliTime() const
 {
-    FastOS_Time time;
-    time.SetNow();
-    return (uint64_t)time.MilliSecs();
+    return duration_cast<milliseconds>(steady_clock::now().time_since_epoch()).count();
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/systemtimer.cpp]: SteadyTimer

File: searchcore/src/vespa/searchcore/proton/server/transactionlogmanagerbase.h (hunk 1 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/transactionlogmanagerbase.h]: Fixed

File: searchcore/src/vespa/searchcore/proton/server/transactionlogmanagerbase.h (hunk 2 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/transactionlogmanagerbase.h]: Fixed

File: searchlib/src/vespa/searchlib/features/random_normal_feature.cpp (hunk 3 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/features/random_normal_feature.cpp]: System time it is.

File: searchlib/src/vespa/searchlib/features/randomfeature.cpp (hunk 4 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/features/randomfeature.cpp]: System time it is.

File: searchlib/src/vespa/searchlib/features/debug_wait.cpp (hunk 6 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/features/debug_wait.cpp]: Fixed

File: searchlib/src/vespa/searchlib/features/debug_wait.cpp (hunk 7 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/features/debug_wait.cpp]: I agree, but I will leave it here until I get a better place.

**Comment by baldersheim** [comment]
Superceeded by #11363, hence closing

**Pull request closed without merging**